### ISSUE TITLE
conda update conda fails with "RemoveError" for setuptools

### ISSUE THREAD

**Issue by jsolbrig** [conda/conda#9337]
## Current Behavior

I am not 100% sure that this is a conda issue, but I am fairly confident that it is.  If it isn't maybe someone here will have an idea on which package is causing the problem.

I am building docker image using the `continuumio/miniconda3:4.7.10` image as my base which [hasn't been changed in two months](https://hub.docker.com/r/continuumio/miniconda3/tags). 
 Yesterday `conda update conda` worked fine, but today it is failing with `RemoveError: 'setuptools' is a dependency of conda and cannot be removed from
conda's operating environment.`.  I don't know why conda has suddenly decided it needs to remove `setuptools`.  My Dockerfile is the same as yesterday and am using the same base image, which also has not changed.  This leads me to believe that it is a conda issue of some sort.

That said, I see that no new release of conda has been created.

### Steps to Reproduce

Below are two docker files.  The first attempts to remove `setuptools` for some reason and fails with `RemoveError: 'setuptools' is a dependency of conda and cannot be removed from
conda's operating environment.`  The second removes `setuptools` with `pip remove -y setuptools`, reinstalls it with `conda install setuptools`, then successfully runs `conda update conda`.

### Broken Dockerfile
```
FROM continuumio/miniconda3:4.7.10 as base

WORKDIR /app

RUN conda update conda
```

#### Output from `docker build -f --no-cache Dockerfile.broken .`
```
> docker build -f Dockerfile.broken .
Sending build context to Docker daemon  3.072kB
Step 1/3 : FROM continuumio/miniconda3:4.7.10 as base
 ---> 4a51de2367be
Step 2/3 : WORKDIR /app
 ---> Using cache
 ---> d5da6c69cd74
Step 3/3 : RUN conda update conda
 ---> Running in 79715e115fdb
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /opt/conda

  added / updated specs:
    - conda


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    asn1crypto-1.0.1           |           py37_0         161 KB
    ca-certificates-2019.8.28  |                0         132 KB
    certifi-2019.9.11          |           py37_0         154 KB
    chardet-3.0.4              |        py37_1003         174 KB
    conda-4.7.12               |           py37_0         3.0 MB
    conda-package-handling-1.6.0|   py37h7b6447c_0         797 KB
    openssl-1.1.1d             |       h7b6447c_2         2.5 MB
    pysocks-1.7.1              |           py37_0          30 KB
    sqlite-3.30.0              |       h7b6447c_0         1.9 MB
    tqdm-4.36.1                |             py_0          50 KB
    ------------------------------------------------------------
                                           Total:         8.9 MB

The following packages will be REMOVED:

  bzip2-1.0.8-h7b6447c_0
  libarchive-3.3.3-h5d8350f_5
  libxml2-2.9.9-hea5a465_1
  lz4-c-1.8.1.2-h14c3975_0
  lzo-2.10-h49e0be7_2
  pip-19.1.1-py37_0
  python-libarchive-c-2.8-py37_11
  setuptools-41.0.1-py37_0
  wheel-0.33.4-py37_0
  zstd-1.3.7-h0b5b093_0

The following packages will be UPDATED:

  asn1crypto                                  0.24.0-py37_0 --> 1.0.1-py37_0
  ca-certificates                               2019.5.15-0 --> 2019.8.28-0
  certifi                                  2019.6.16-py37_0 --> 2019.9.11-py37_0
  chardet                                      3.0.4-py37_1 --> 3.0.4-py37_1003
  conda                                       4.7.10-py37_0 --> 4.7.12-py37_0
  conda-package-han~                          1.3.11-py37_0 --> 1.6.0-py37h7b6447c_0
  openssl                                 1.1.1c-h7b6447c_1 --> 1.1.1d-h7b6447c_2
  pysocks                                      1.7.0-py37_0 --> 1.7.1-py37_0
  sqlite                                  3.29.0-h7b6447c_0 --> 3.30.0-h7b6447c_0
  tqdm                                          4.32.1-py_0 --> 4.36.1-py_0


Proceed ([y]/n)?

Downloading and Extracting Packages
chardet-3.0.4        | 174 KB    | ########## | 100%
openssl-1.1.1d       | 2.5 MB    | ########## | 100%
tqdm-4.36.1          | 50 KB     | ########## | 100%
ca-certificates-2019 | 132 KB    | ########## | 100%
conda-4.7.12         | 3.0 MB    | ########## | 100%
conda-package-handli | 797 KB    | ########## | 100%
asn1crypto-1.0.1     | 161 KB    | ########## | 100%
pysocks-1.7.1        | 30 KB     | ########## | 100%
certifi-2019.9.11    | 154 KB    | ########## | 100%
sqlite-3.30.0        | 1.9 MB    | ########## | 100%
Preparing transaction: ...working... done
Verifying transaction: ...working... failed

RemoveError: 'setuptools' is a dependency of conda and cannot be removed from
conda's operating environment.

The command '/bin/sh -c conda update conda' returned a non-zero code: 1
```

### Working Dockerfile
```
FROM continuumio/miniconda3:4.7.10 as base

WORKDIR /app

RUN pip uninstall -y setuptools && \
    conda install setuptools && \
    conda update conda
```

#### Output from `docker build --no-cache -f Dockerfile.working .`
```
> docker build --no-cache -f Dockerfile.working .
Sending build context to Docker daemon  3.072kB
Step 1/3 : FROM continuumio/miniconda3:4.7.10 as base
 ---> 4a51de2367be
Step 2/3 : WORKDIR /app
 ---> Running in b0c3333f3b0a
Removing intermediate container b0c3333f3b0a
 ---> 9a133cb409c1
Step 3/3 : RUN pip uninstall -y setuptools &&     conda install setuptools &&     conda update conda
 ---> Running in 6478847837bc
Uninstalling setuptools-41.0.1:
  Successfully uninstalled setuptools-41.0.1
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /opt/conda

  added / updated specs:
    - setuptools


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2019.8.28  |                0         132 KB
    certifi-2019.9.11          |           py37_0         154 KB
    conda-4.7.12               |           py37_0         3.0 MB
    openssl-1.1.1d             |       h7b6447c_2         2.5 MB
    setuptools-41.4.0          |           py37_0         651 KB
    ------------------------------------------------------------
                                           Total:         6.4 MB

The following packages will be UPDATED:

  ca-certificates                               2019.5.15-0 --> 2019.8.28-0
  certifi                                  2019.6.16-py37_0 --> 2019.9.11-py37_0
  conda                                       4.7.10-py37_0 --> 4.7.12-py37_0
  openssl                                 1.1.1c-h7b6447c_1 --> 1.1.1d-h7b6447c_2
  setuptools                                  41.0.1-py37_0 --> 41.4.0-py37_0


Proceed ([y]/n)?

Downloading and Extracting Packages
ca-certificates-2019 | 132 KB    | ########## | 100%
openssl-1.1.1d       | 2.5 MB    | ########## | 100%
conda-4.7.12         | 3.0 MB    | ########## | 100%
certifi-2019.9.11    | 154 KB    | ########## | 100%
setuptools-41.4.0    | 651 KB    | ########## | 100%
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /opt/conda

  added / updated specs:
    - conda


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    asn1crypto-1.0.1           |           py37_0         161 KB
    chardet-3.0.4              |        py37_1003         174 KB
    conda-package-handling-1.6.0|   py37h7b6447c_0         797 KB
    pysocks-1.7.1              |           py37_0          30 KB
    sqlite-3.30.0              |       h7b6447c_0         1.9 MB
    tqdm-4.36.1                |             py_0          50 KB
    ------------------------------------------------------------
                                           Total:         3.1 MB

The following packages will be REMOVED:

  bzip2-1.0.8-h7b6447c_0
  libarchive-3.3.3-h5d8350f_5
  libxml2-2.9.9-hea5a465_1
  lz4-c-1.8.1.2-h14c3975_0
  lzo-2.10-h49e0be7_2
  pip-19.1.1-py37_0
  python-libarchive-c-2.8-py37_11
  wheel-0.33.4-py37_0
  zstd-1.3.7-h0b5b093_0

The following packages will be UPDATED:

  asn1crypto                                  0.24.0-py37_0 --> 1.0.1-py37_0
  chardet                                      3.0.4-py37_1 --> 3.0.4-py37_1003
  conda-package-han~                          1.3.11-py37_0 --> 1.6.0-py37h7b6447c_0
  pysocks                                      1.7.0-py37_0 --> 1.7.1-py37_0
  sqlite                                  3.29.0-h7b6447c_0 --> 3.30.0-h7b6447c_0
  tqdm                                          4.32.1-py_0 --> 4.36.1-py_0


Proceed ([y]/n)?

Downloading and Extracting Packages
tqdm-4.36.1          | 50 KB     | ########## | 100%
sqlite-3.30.0        | 1.9 MB    | ########## | 100%
conda-package-handli | 797 KB    | ########## | 100%
asn1crypto-1.0.1     | 161 KB    | ########## | 100%
pysocks-1.7.1        | 30 KB     | ########## | 100%
chardet-3.0.4        | 174 KB    | ########## | 100%
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Removing intermediate container 6478847837bc
 ---> 6e7c3777bf3d
Successfully built 6e7c3777bf3d
```

## Expected Behavior

Conda shouldn't try to uninstall setuptools.

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
     active environment : None
       user config file : /root/.condarc
 populated config files :
          conda version : 4.7.10
    conda-build version : not installed
         python version : 3.7.3.final.0
       virtual packages :
       base environment : /opt/conda  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /opt/conda/pkgs
                          /root/.conda/pkgs
       envs directories : /opt/conda/envs
                          /root/.conda/envs
               platform : linux-64
             user-agent : conda/4.7.10 requests/2.22.0 CPython/3.7.3 Linux/4.9.184-linuxkit debian/10 glibc/2.28
                UID:GID : 0:0
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
Reported Nothing
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
# packages in environment at /opt/conda:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
asn1crypto                0.24.0                   py37_0    defaults
bzip2                     1.0.8                h7b6447c_0    defaults
ca-certificates           2019.5.15                     0    defaults
certifi                   2019.6.16                py37_0    defaults
cffi                      1.12.3           py37h2e261b9_0    defaults
chardet                   3.0.4                    py37_1    defaults
conda                     4.7.10                   py37_0    defaults
conda-package-handling    1.3.11                   py37_0    defaults
cryptography              2.7              py37h1ba5d50_0    defaults
idna                      2.8                      py37_0    defaults
libarchive                3.3.3                h5d8350f_5    defaults
libedit                   3.1.20181209         hc058e9b_0    defaults
libffi                    3.2.1                hd88cf55_4    defaults
libgcc-ng                 9.1.0                hdf63c60_0    defaults
libstdcxx-ng              9.1.0                hdf63c60_0    defaults
libxml2                   2.9.9                hea5a465_1    defaults
lz4-c                     1.8.1.2              h14c3975_0    defaults
lzo                       2.10                 h49e0be7_2    defaults
ncurses                   6.1                  he6710b0_1    defaults
openssl                   1.1.1c               h7b6447c_1    defaults
pip                       19.1.1                   py37_0    defaults
pycosat                   0.6.3            py37h14c3975_0    defaults
pycparser                 2.19                     py37_0    defaults
pyopenssl                 19.0.0                   py37_0    defaults
pysocks                   1.7.0                    py37_0    defaults
python                    3.7.3                h0371630_0    defaults
python-libarchive-c       2.8                     py37_11    defaults
readline                  7.0                  h7b6447c_5    defaults
requests                  2.22.0                   py37_0    defaults
ruamel_yaml               0.15.46          py37h14c3975_0    defaults
setuptools                41.0.1                   py37_0    defaults
six                       1.12.0                   py37_0    defaults
sqlite                    3.29.0               h7b6447c_0    defaults
tk                        8.6.8                hbc83047_0    defaults
tqdm                      4.32.1                     py_0    defaults
urllib3                   1.24.2                   py37_0    defaults
wheel                     0.33.4                   py37_0    defaults
xz                        5.2.4                h14c3975_4    defaults
yaml                      0.1.7                had09818_2    defaults
zlib                      1.2.11               h7b6447c_3    defaults
zstd                      1.3.7                h0b5b093_0    defaults
```
</p></details>

**Comment by langevin-usgs** [comment]
We are seeing this same problem in our Travis testing with Miniconda: "conda update conda" tries to remove setuptools.

```
$ conda update -q conda
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done
## Package Plan ##
  environment location: /home/travis/miniconda
  added / updated specs:
    - conda
The following packages will be downloaded:
    package                    |            build
    ---------------------------|-----------------
    asn1crypto-1.0.1           |           py37_0         161 KB
    ca-certificates-2019.8.28  |                0         132 KB
    certifi-2019.9.11          |           py37_0         154 KB
    chardet-3.0.4              |        py37_1003         174 KB
    conda-4.7.12               |           py37_0         3.0 MB
    conda-package-handling-1.6.0|   py37h7b6447c_0         797 KB
    openssl-1.1.1d             |       h7b6447c_2         2.5 MB
    pysocks-1.7.1              |           py37_0          30 KB
    sqlite-3.30.0              |       h7b6447c_0         1.9 MB
    tqdm-4.36.1                |             py_0          50 KB
    ------------------------------------------------------------
                                           Total:         8.9 MB
The following packages will be REMOVED:
  bzip2-1.0.8-h7b6447c_0
  libarchive-3.3.3-h5d8350f_5
  libxml2-2.9.9-hea5a465_1
  lz4-c-1.8.1.2-h14c3975_0
  lzo-2.10-h49e0be7_2
  pip-19.1.1-py37_0
  python-libarchive-c-2.8-py37_11
  setuptools-41.0.1-py37_0
  wheel-0.33.4-py37_0
  zstd-1.3.7-h0b5b093_0
The following packages will be UPDATED:
  asn1crypto                                  0.24.0-py37_0 --> 1.0.1-py37_0
  ca-certificates                               2019.5.15-0 --> 2019.8.28-0
  certifi                                  2019.6.16-py37_0 --> 2019.9.11-py37_0
  chardet                                      3.0.4-py37_1 --> 3.0.4-py37_1003
  conda                                       4.7.10-py37_0 --> 4.7.12-py37_0
  conda-package-han~                          1.3.11-py37_0 --> 1.6.0-py37h7b6447c_0
  openssl                                 1.1.1c-h7b6447c_1 --> 1.1.1d-h7b6447c_2
  pysocks                                      1.7.0-py37_0 --> 1.7.1-py37_0
  sqlite                                  3.29.0-h7b6447c_0 --> 3.30.0-h7b6447c_0
  tqdm                                          4.32.1-py_0 --> 4.36.1-py_0
Preparing transaction: ...working... done
Verifying transaction: ...working... failed
RemoveError: 'setuptools' is a dependency of conda and cannot be removed from
conda's operating environment.
The command "conda update -q conda" failed and exited with 1 during .
Your build has been stopped.
```

Full log is here: https://travis-ci.org/MODFLOW-USGS/modflow6/jobs/598415765
Travis yml is here: https://github.com/MODFLOW-USGS/modflow6/blob/develop/.travis.yml

The problem just started today (Oct 15).  "conda update conda" worked fine yesterday (Oct 14).

**Comment by jsolbrig** [comment]
@langevin-usgs Just in case you didn't see, there is a workaround in my initial post.

```
pip uninstall -y setuptools
conda install setuptools
conda update conda
```

But that shouldn't need to be done.

**Comment by msarahan** [comment]
Seems like a metadata issue, since conda itself hasn't changed. Several of the removed packages are dependencies if probably itself. It is very odd to see them being removed if python itself is not seemingly participating.

The most likely explanation given your fix is that a metadata hotfix has made setuptools in the 4.7.10 installer inconsistent somehow. Inconsistencies lead quickly to bizarre issues like this.

For a faster, more robust way to bootstrap your environment, see https://github.com/conda/conda-prefix-replacement/blob/master/.travis.yml#L9-L14

We need to fix this, but the standalone approach is better anyway.

**Comment by eric-wieser** [comment]
@msarahan

> For a faster, more robust way to bootstrap your environment, see

Is there an equivalent way you would recommend for linux? I'm following the instructions [here](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/use-conda-with-travis-ci.html), and running into the same problem.

**Comment by msarahan** [comment]
We're still not sure why this problem only suddenly cropped up. We think it is addressed in #9014, which was part of 4.7.11.

New miniconda was released today with this fix. Hopefully these issues will disappear.

**Comment by Bocha84** [comment]
> We're still not sure why this problem only suddenly cropped up. We think it is addressed in #9014, which was part of 4.7.11.
> 
> New miniconda was released today with this fix. Hopefully these issues will disappear.

We saw this issue with miniconda 4.7.12

**Comment by msarahan** [comment]
Please link to a CI build that shows the problem.

**Comment by YoungPyDawan** [comment]
> Please link to a CI build that shows the problem.

https://hub.docker.com/layers/continuumio/miniconda3/4.7.10/images/sha256-54eb3dd4003f11f6a651b55fc2074a0ed6d9eeaa642f1c4c9a7cf8b148a30ceb

**Comment by msarahan** [comment]
@YoungPyDawan please do not delete comments. It makes discussions really hard to follow. What you link to there is our docker image, not a CI log. If you don't use CI, please make sure that you are using it latest docker image (use docker pull as necessary to update it).  CI logs are nice because they are a public build log. If you don't have that, you need to provide your log a different way. Either copy/paste it upload it as a text file here.

**Comment by dgdekoning** [comment]
> Seems like a metadata issue, since conda itself hasn't changed. Several of the removed packages are dependencies if probably itself. It is very odd to see them being removed if python itself is not seemingly participating.
> 
> The most likely explanation given your fix is that a metadata hotfix has made setuptools in the 4.7.10 installer inconsistent somehow. Inconsistencies lead quickly to bizarre issues like this.
> 
> For a faster, more robust way to bootstrap your environment, see https://github.com/conda/conda-prefix-replacement/blob/master/.travis.yml#L9-L14
> 
> We need to fix this, but the standalone approach is better anyway.

Thanks @msarahan! I went and implemented your suggested standalone approach and it works like a dream.

**Comment by jsolbrig** [comment]
@msarahan I just got back to this issue, but it does appear that something (maybe #9014) fixed the problem.  I'm able to run my CI correctly using both 4.7.11 and 4.7.12.  Should something be done to either patch 4.7.10 with #9014 or indicate somewhere that 4.7.10 is broken and should not be used?

**Comment by jjhelmus** [comment]
@jsolbrig The dependency information for conda 4.7.10 and other versions has been modified to exclude `setuptools` from the `run_constrained` section which should address this issue.

**Comment by jsolbrig** [comment]
@jjhelmus Thank you, that does appear to solve the problem!

**Comment by somexgupta** [comment]
> @langevin-usgs Just in case you didn't see, there is a workaround in my initial post.
> 
> ```
> pip uninstall -y setuptools
> conda install setuptools
> conda update conda
> ```
> 
> But that shouldn't need to be done.

pip uninstall -y setuptools
#conda install setuptools -> was throwing an error because we where modifying from conda environment
pip install setuptools # this worked for me to install the latest package of setuptools
conda update -all
